mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -49,152 +49,157 @@ def _autolog(log_every_n_epoch=1, log_models=True):
+                    metrics = dict(map(lambda x: (x[0], float(x[1])), cur_metrics.items()))
+                    metrics_logger.record_metrics(metrics, pl_module.current_epoch)
+
+                for callback in trainer.callbacks:
+                    if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                        self._early_stop_check(callback)
+
+            def on_train_start(self, trainer, pl_module):
+                """
+                Logs Optimizer related metrics when the train begins
+
+                :param trainer: pytorch lightning trainer instance
+                :param pl_module: pytorch lightning base module
+                """
+                try_mlflow_log(mlflow.set_tag, "Mode", "training")
+                try_mlflow_log(mlflow.log_param, "epochs", trainer.max_epochs)
+
+                for callback in trainer.callbacks:
+                    if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                        self.early_stopping = True
+                        self._log_early_stop_params(callback)
+
+                # TODO For logging optimizer params - Following scenarios are to revisited.
+                # 1. In the current scenario, only the first optimizer details are logged.
+                #    Code to be enhanced to log params when multiple optimizers are used.
+                # 2. mlflow.log_params is used to store optimizer default values into mlflow.
+                #    The keys in default dictionary are too short, Ex: (lr - learning_rate).
+                #    Efficient mapping technique needs to be introduced
+                #    to rename the optimizer parameters based on keys in default dictionary.
+
+                if hasattr(trainer, "optimizers"):
+                    optimizer = trainer.optimizers[0]
+                    try_mlflow_log(mlflow.log_param, "optimizer_name", type(optimizer).__name__)
+
+                    if hasattr(optimizer, "defaults"):
+                        try_mlflow_log(mlflow.log_params, optimizer.defaults)
+
+                summary = str(ModelSummary(pl_module, mode="full"))
+                tempdir = tempfile.mkdtemp()
+                try:
+                    summary_file = os.path.join(tempdir, "model_summary.txt")
+                    with open(summary_file, "w") as f:
+                        f.write(summary)
+
+                    try_mlflow_log(mlflow.log_artifact, local_path=summary_file)
+                finally:
+                    shutil.rmtree(tempdir)
+
+            def on_train_end(self, trainer, pl_module):
+                """
+                Logs the model checkpoint into mlflow - models folder on the training end
+
+                :param trainer: pytorch lightning trainer instance
+                :param pl_module: pytorch lightning base module
+                """
+                # manually flushing any remaining metrics from training.
+                metrics_logger._flush()
```

Comment:
Can we make this a public API on `BatchMetricsLogger` (i.e. drop the `_`) since we have legitimate use cases for calling it outside of the module where it's defined?

Reply:
@dbczumar  Now that I made it a public API do you think I need to add a unti test for `flush()` in `test_autologging_utils.py`?